src/Variant.cpp

```diff
@@ -83,33 +83,347 @@
+  std::istringstream stream(str);
+  std::vector<T> vec;
+  T val;
+  while (stream >> val) {
+    vec.emplace_back(val);
+    if ((stream >> std::ws).peek() == ',') {
+      stream.ignore();
+    }
+  }
+  return vec;
+}
+
+// creates a vector of parsed std::string from split string(len=...) pieces
+std::vector<std::string> split_variant_strings (const std::string &str) {
+  std::vector<std::string> strings;
+  const std::string begin("string(len=");
+  const std::string mid(", val=\"");
+  const std::string end("\")");
+
+  size_t current_idx = 0;
+  while ((current_idx = str.find(begin, current_idx))
+          != std::string::npos)
+  {
+    std::string lenstring;
+    for (int i = current_idx + begin.size(); std::isdigit(str[i]); i++) {
+      lenstring.push_back(str[i]);
+    }
+    if (lenstring.size() == 0) {
+      throw std::invalid_argument("Not a valid string variant string, "
+                                  "must specify a string length");
+    }
+    long length = std::stol(lenstring);
+    if (length < 0) {
+      throw std::invalid_argument("Not a valid string variant string, "
+                                  "string length must be non-negative");
+    }
+    if (!substrEquals(str, current_idx + begin.size() + lenstring.size(),
+                      mid))
+    {
+      throw std::invalid_argument("Not a valid string variant string, "
+                                  "mid pattern expected ('" + mid + "')");
+    }
+    if (!substrEquals(
+           str,
+           current_idx + begin.size() + lenstring.size() + mid.size() + length,
+           end))
+    {
+      throw std::invalid_argument("Not a valid string variant string, "
+                                  "ending pattern expected ('" + end + "')");
+    }
+    if (str.size() <= current_idx + begin.size() + lenstring.size()
+                      + mid.size() + length + end.size())
+    {
+      throw std::invalid_argument("Not a valid string variant string, "
+                                  "value ends prematurely");
+    }
+    auto val_begin = str.begin() + current_idx + begin.size()
+                     + lenstring.size() + mid.size();
+    strings.emplace_back(val_begin, val_begin + length);
+    current_idx += length;
+  }
+  return strings;
+}
+
+} // end of unnamed namespace
+
 namespace flit {
 
-std::ostream& operator<< (std::ostream& out, const Variant &val) {
-  switch (val.type()) {
+template <> long double Variant::val() const {
+  return this->longDouble();
+}
+
+template <> std::string Variant::val() const {
+  return this->string();
+}
+
+template <> std::vector<std::string> Variant::val() const {
+  return this->vectorString();
+}
+
+template <> std::vector<float> Variant::val() const {
+  return this->vectorFloat();
+}
+
+template <> std::vector<double> Variant::val() const {
+  return this->vectorDouble();
+}
+
+template <> std::vector<long double> Variant::val() const {
+  return this->vectorLongDouble();
+}
```

Comment:
All of these above `val()` variants can be inline functions in the header file.  That would probably be better.